rfcs/0000-operator-managed-apim-config.md

```diff
@@ -0,0 +1,223 @@
+    }
+}
+
+Person(consumer, "API consumer")
+
+BiRel(crd1, operator, "Synchronize Api product 1")
+BiRel(crd2, operator, "Synchronize Api product 2")
+BiRel(operator, product1, "Synchronize Api product 1")
+BiRel(operator, product2, "Synchronize Api product 2")
+Rel(consumer, product1, "Consumes")
+Rel(consumer, product2, "Consumes")
+UpdateLayoutConfig($c4ShapeInRow="3", $c4BoundaryInRow="1")
+```
+
+Sequence diagram of create/upgrade scenario
+```mermaid
+sequenceDiagram
+participant dev as Developer
+participant kapi as Kubernetes API
+participant ctrl as Operator
+participant apim as Api Management
+dev ->> kapi: Creates/Update ProductApi specification
+kapi ->> ctrl: Notifies about new/changed ProductAPI
+ctrl ->> kapi: Read desired ProductApi specification
+alt if actual state does not match desired state
+  ctrl ->> apim: Read actual state in Api management
+  ctrl ->> apim: Update state in Api management to match desired state
+  ctrl ->> apim: Check status of update job
+  ctrl ->> kapi: Update status of ProductApi
+end
+```
+
+Sequence diagram sync loop
+```mermaid
+sequenceDiagram
+participant dev as Developer
+participant kapi as Kubernetes API
+participant ctrl as Operator
+participant apim as Api Management
+kapi ->> ctrl: Reconcile of ProductApi
+ctrl ->> kapi: Read desired ProductApi specification
+ctrl ->> apim: Read actual state in Api management
+alt if actual state does not match desired state
+  ctrl ->> apim: Update state in Api management to match desired state
+  ctrl ->> apim: Check status of update job
+  ctrl ->> kapi: Update status of ProductApi
+end
+```
+
+Sequence diagram status api
+```mermaid
+sequenceDiagram
+participant dev as Developer
+participant api as StatusApi (custom)
+participant kapi as Kubernetes API
+dev ->> api: Get status for deployment
+api ->> kapi: Get status of resouce in kubernetes
+kapi ->> api: Returns current status of resource(s)
+api ->> dev: Returns current status of deployment
+```
+
+# Drawbacks
+
+* We need to learn and understand the kubernetes operator pattern
+* Tied to kubernetes as a "provider"/"system"
+* Faults in the operator will affect all the APIs it manages
+
+# Rationale and alternatives
+
+* Terraform modules that each team deploys
+* Bicep modules that each team deploys
+* Centralized IaC configuration
+* Scripts and pipelines
+* Azure service operator
+
+# Prior art
+
+We have a pipeline script setup today that configures the api management service. This is complex and has alot of configuration elements that most of the developers shouldn't have to care about.
+
+Few people know how this works and it is complicated to deploy a new or a change to a api in the api management service
+
+# Unresolved questions
+
+* Can we automate even more of the process than we do today
+* How we should report status of an deployment back to the teams
+  * Flux status?
+  * Custom api endpoint for viewing status (ala. kuberneteswrapper)
```

Comment:
Havn't given it to much thought, only that the teams deploying/configuring apis need a, ideally single, pane of glass for viewing the current status of the deployment/config.
I'm not sure flux can provide us with the functionality we need (eg. waiting for a given status of our crd).
As we are planning to deploy this operator to a management cluster and this has access to APIM we might not want to expose that cluster to the internet in any way, so we might want to add a service that can access the management cluster and fetch the status and other relevant metadata (failure messages, deployed version, etc).
Does that make it clearer? if so I can try to add the context into the rfc future works

Reply:
> As we are planning to deploy this operator to a management cluster

Ahh, I didn't know this, thought it was gonna run in "user" clusters - couldn't find that in the RFC. Flux can only give you the state of the deployment of the artifact, so that means either **deployed new version** or **rollbacked because of Err("some err")**, but thay only applies to the CR. I guess we can rely on the controller to provide that info for errors and expose that in the status of the CR?

 Exposing it to the public should be fine? It's the same as what happens currently with an App (in practice either a deployment or daemonset, but let's assume we're already using App):

-  There's a CR for an App which gets it's current state exposed (version, name, etc) - this is the **current state**
- There's a **desired state**, which lives stored somewhere as a part of an artifact

Disctl can handle now the **current state**, and it should be fairly easy to display also the **desired state**. If you replace the App CR with this Apim CR, then the process is identical and it doesn't access to the management cluster.

Reply:
Yeah I think we are thinking on the same solution. Using disctl is a nice idea! I can add in a comment to say that we plan to deploy it in i a management cluster. If we decide to change that we can come back and create a new version of this RFC-document with the new deployment strategy

Reply:
Added a sentence about deployment to a management cluster and wrote up that we need to use flux to report deploy time failures, but as deployment is asynchronous we also need a api where status can be checked.
Disctl can the be extended to leverage this api and display the status